File: app/src/main/java/com/robotnec/budget/core/persistence/TransactionContextImpl.java
package com.robotnec.budget.core.persistence;

import com.yahoo.squidb.data.SquidDatabase;

/**
 * @author zak <zak@swingpulse.com>
 */
public class TransactionContextImpl implements TransactionContext {

    private final SquidDatabase database;

    public TransactionContextImpl(SquidDatabase database) {
        this.database = database;
    }

    @Override
    public boolean doInTransaction(Transaction transaction) {
        boolean success = false;
        database.beginTransaction();
        try {
            success = transaction.execute();
            if (success) {
                database.setTransactionSuccessful();
            }
        } finally {
            database.endTransaction();
        }
        return success;
    }
}

File: app/src/main/java/com/robotnec/budget/core/persistence/TransactionContextImpl.kt
package com.robotnec.budget.core.persistence

import com.yahoo.squidb.data.SquidDatabase

/**
 * @author zak zak@swingpulse.com>
 */
class TransactionContextImpl(private val database: SquidDatabase) : TransactionContext {

    override fun doInTransaction(transaction: () -> Boolean): Boolean {
        val success: Boolean
        database.beginTransaction()
        try {
            success = transaction()
            if (success) {
                database.setTransactionSuccessful()
            }
        } finally {
            database.endTransaction()
        }
        return success
    }
}
